http/resource.go

```diff
@@ -180,6 +180,17 @@ var resourcePatchHandler = withUser(func(w http.ResponseWriter, r *http.Request,
 			if !d.user.Perm.Create {
 				return errors.ErrPermissionDenied
 			}
+
+			rel, err := filepath.Rel(src, dst)//nolint:shadow
+			if err != nil {
+				return err
+			}
+
+			rel = filepath.ToSlash(rel)
+			if !strings.HasPrefix(rel, "../") && rel != ".." {
+				return errors.ErrSourceIsParent
+			}
```

Comment:
don't we want to do the same check for `rename`?

Reply:
Yeah, this is really necessary. I can add a function for checking:
```go
func CheckParent(src, dst string) error {
	rel, err := filepath.Rel(src, dst)
	if err != nil {
		return err
	}

	rel = filepath.ToSlash(rel)
	if !strings.HasPrefix(rel, "../") && rel != ".." {
		return errors.ErrSourceIsParent
	}

	return nil
}
```
And call it after permission checking on copy and rename actions:
```go
err = CheckParent(src, dst)
if err != nil {
	return err
}
```

Reply:
Looks good

Reply:
@o1egl for consistency, i am thinking to refactor the renaming into a function too:
```go
func AddSuffix(path string, fs afero.Fs) string {
	counter := 1
	dir, name := filepath.Split(path)
	ext := filepath.Ext(name)
	base := strings.TrimSuffix(name, ext)

	for {
		if _, err := fs.Stat(path); err != nil {
			break
		}
		renamed := fmt.Sprintf("%s(%d)%s", base, counter, ext)
		path = filepath.ToSlash(dir) + renamed
		counter++
	}

	return path
}
```
So it will be called after permission checking on copy and rename:
```go
if rename {
	dst = AddSuffix(dst, d.user.Fs)
}
```
Can i amend that into the parent verification commit?

Reply:
Sure. But maybe `addVersionSuffix` or something like that will give more context of that function.